llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2538,6 +2538,50 @@ LogicalResult AttentionOp::reifyResultShapes(
       .reifyResultShapes(b, reifiedReturnShapes);
 }
 
+//===----------------------------------------------------------------------===//
+// TiledAttentionOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult TiledAttentionOp::verify() {
+  Operation *op = getOperation();
+  ShapedType queryType = getQueryType();
+  ShapedType keyType = getKeyType();
+  ShapedType valueType = getValueType();
+  ShapedType tiledResultType = getTiledResultType();
+  ShapedType maxType = getMaxType();
+  ShapedType sumType = getSumType();
+  ArrayRef<int64_t> queryShape = queryType.getShape();
+  ArrayRef<int64_t> keyShape = keyType.getShape();
+  ArrayRef<int64_t> valueShape = valueType.getShape();
+  ArrayRef<int64_t> tiledResultShape = tiledResultType.getShape();
+  ArrayRef<int64_t> maxShape = maxType.getShape();
+  ArrayRef<int64_t> sumShape = sumType.getShape();
+  if (failed(verifyCompatibleShape(keyShape, valueShape)))
+    return op->emitOpError("incompatible value shape");
+  if (failed(verifyCompatibleShape(queryShape, tiledResultShape)))
+    return op->emitOpError("incompatible tiled result shape");
+  if (failed(verifyCompatibleShape(maxShape, sumShape)))
+    return op->emitOpError("incompatible sum shape");
+  if (keyShape[0] != queryShape[0])
```

Comment:
If I understand this right, this op is doing batch tiling? So the tensors go from B x T/S x d to T/S x d? For self-attention, the dim 0 of query, key and value will match. But for cross-attention, dim 0 of key and value should match and dim 0 of query and result should match. So shouldn't this check be comparing queryShape[0] and valueShape[0] ?

Reply:
With the latest push, I've removed this new op and have instead adapted the vanilla attention to instead adapt as per the resulting tensors.
I see that the check `keyShape[0] != queryShape[0]` is [already part of the vanilla attention](https://github.com/openxla/iree/blob/main/llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp#L2434-L2435) on ToM - and since the "tiled version" would essentially form via a valid attention op, I'm currently not performing this check for FlashAttention.

But let me know if we'd need to address this particular check in vanilla attention's `verify` function in some way.
To me the particular check we're talking about seems correct - please let me know where I'm going wrong in my understanding.
I'm trying to follow the logical statements you mentioned :-
```
For cross-attention :-
   1. dim-0 of KEY == dim-0 VALUE
   2. dim-0 of QUERY == dim-0 OUTPUT
Therefore, required check is => dim-0 QUERY == dim-0 of VALUE ? 
```
If we indeed need `dim-0 QUERY == dim-0 of VALUE` then the above check is logically equivalent since `dim-0 of KEY == dim-0 VALUE`.
But if we need `dim-0 QUERY == dim-0 of OUTPUT`, then that is [already being verified](https://github.com/openxla/iree/blob/main/llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp#L2432-L2433).

Reply:
Just adding here as a clarification to maintain the discussion thread.

As @harsh-nod mentioned earlier we only want to ensure two checks :-
1. `dim-0 of KEY == dim-0 VALUE`
2. `dim-0 of QUERY == dim-0 OUTPUT`

This therefore means the check `keyShape[0] != queryShape[0]` should essentially just go away!

@harsh-nod - thanks for clarifying this!

I've addressed this in the recent push - you may take a look. :)